### PR TITLE
Add nil check for billing mode in AWS DynamoDB events driver

### LINKED ISSUE #12443: Teleport 9.2.0 panics with DynamoDB
The logic for checking the billing mode is missing some nil checks:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x3705540]

goroutine 1 [running, locked to thread]:
github.com/gravitational/teleport/lib/events/dynamoevents.(*Log).getBillingModeIsProvisioned(0xc000b6c0e0, {0x5c522f8, 0xc000c9a880})
	/go/src/github.com/gravitational/teleport/lib/events/dynamoevents/dynamoevents.go:1177 +0xc0
github.com/gravitational/teleport/lib/events/dynamoevents.New({0x5c522f8, 0xc000c9a880}, {{0xc00007d570, 0xa}, {0xc0004ce2db, 0x25}, 0xa, 0xa, 0xc0002f8de0, {0x5c74978, ...}, ...}, ...)
	/go/src/github.com/gravitational/teleport/lib/events/dynamoevents/dynamoevents.go:307 +0x56c
github.com/gravitational/teleport/lib/service.initExternalLog({0x5c522f8, 0xc000c9a880}, {0x5d02670, 0xc000653340}, {0xc000b2a4c8, 0x15}, {0x5cd9340, 0xc00083ca20})
	/go/src/github.com/gravitational/teleport/lib/service/service.go:1078 +0x878
github.com/gravitational/teleport/lib/service.(*TeleportProcess).initAuthService(0xc000c7dcc0)
	/go/src/github.com/gravitational/teleport/lib/service/service.go:1183 +0x3c5
github.com/gravitational/teleport/lib/service.NewTeleport(0xc000293000)
	/go/src/github.com/gravitational/teleport/lib/service/service.go:807 +0x1572
github.com/gravitational/teleport/e/tool/teleport/process.NewTeleport(0xc000293000)
	/go/src/github.com/gravitational/teleport/e/tool/teleport/process/process.go:67 +0x3b1
github.com/gravitational/teleport/lib/service.Run({_, _}, {{0xc00007c3c0, 0x2}, {0x4aade76, 0x11}, {0xc0004ce0c0, 0x30}, {0x0, 0x0}, ...}, ...)
	/go/src/github.com/gravitational/teleport/lib/service/service.go:506 +0xba
main.main()
	/go/src/github.com/gravitational/teleport/e/tool/teleport/main.go:27 +0x26f
```

### PR DESCRIPTION
**Pull request by xacrimon**
Turns out, AWS is inconsistent in what fields they set and they *don't always*  set the `BillingModeSummary` field if the table is in provisioned mode. Wasn't able to find that noted anywhere so I assumed it would be, this turned out to be incorrect.

This PR adds some pessimistic null-checks and assumed the billing mode is provisioned if the fields are not set. Still not sure *when*  they are set, just that they *sometimes aren't*. This was further not helped by our AWS tests somehow not failing on a panic, they just seem to pause and continue? That's for another PR though.

This also fixes a panic case in `dynamic.go` if the provided map is lacking an event type key which caused AWS tests to fail.

Fix #12443

### CODE REVIEW

**Comment by Tener** [comment]
It would be good to have some tests along with these changes.

**Comment by Tener** [comment]
From failed unit tests:
```
panic: reflect: call of reflect.Value.IsNil on string Value

goroutine 11921 [running]:
reflect.Value.IsNil(...)
	/opt/go/src/reflect/value.go:1440
github.com/gravitational/teleport/lib/events.isInterfaceNil({0x4449fa0, 0xc00422f4b0})
	/workspace/lib/events/dynamic.go:33 +0x25f
github.com/gravitational/teleport/lib/events.FromEventFields(0x5a1fff8)
	/workspace/lib/events/dynamic.go:50 +0xc5
github.com/gravitational/teleport/lib/events.(*FileLog).searchEventsWithFilter(0xc008fc4658, {0xcb17b4, 0x0, 0x0}, {0xc008fc4df8, 0x0, 0x0}, {0x0, 0x75b18c0}, 0x1f4, ...)
	/workspace/lib/events/filelog.go:229 +0x57a
github.com/gravitational/teleport/lib/events.(*FileLog).SearchEvents(0xc0036ea120, {0xc008fc4a88, 0x0, 0x0}, {0x59e25a0, 0xc0076ce480, 0x0}, {0xc00962f049, 0x7}, {0xc005745340, ...}, ...)
	/workspace/lib/events/filelog.go:171 +0x347
github.com/gravitational/teleport/lib/events.(*AuditLog).SearchEvents(0xc001760000, {0x5a584b0, 0xc008ce0360, 0x0}, {0x5a9ee88, 0x4668da0, 0x0}, {0xc00962f049, 0x7}, {0xc005745340, ...}, ...)
	/workspace/lib/events/auditlog.go:874 +0x5eb
github.com/gravitational/teleport/lib/auth.(*ServerWithRoles).SearchEvents(0xc006fc4310, {0x48e6244, 0x2, 0x0}, {0xc001287050, 0x29, 0x0}, {0xc00962f049, 0x7}, {0xc005745340, ...}, ...)
	/workspace/lib/auth/auth_with_roles.go:3760 +0x1b8
github.com/gravitational/teleport/lib/auth.(*GRPCServer).GetEvents(0x0, {0x5a584b0, 0xc008ce0390}, 0xc004bdab40)
	/workspace/lib/auth/grpcserver.go:2867 +0x26c
github.com/gravitational/teleport/api/client/proto._AuthService_GetEvents_Handler.func1({0x5a584b0, 0xc008ce0390}, {0x482b000, 0xc004bdab40})
	/workspace/api/client/proto/authservice.pb.go:16443 +0x8c
github.com/gravitational/teleport/lib/auth.(*Middleware).withAuthenticatedUserUnaryInterceptor(0x5a30e5, {0x5a584b0, 0xc008ce00c0}, {0x482b000, 0xc004bdab40}, 0x9, 0xc009016ca8)
	/workspace/lib/auth/middleware.go:394 +0xc6
github.com/gravitational/teleport/lib/utils.ChainUnaryServerInterceptors.func1.1({0x5a584b0, 0xc008ce00c0}, {0x482b000, 0xc004bdab40})
	/workspace/lib/utils/grpc.go:48 +0x8f
github.com/gravitational/teleport/lib/limiter.(*Limiter).UnaryServerInterceptorWithCustomRate.func1({0x5a584b0, 0xc008ce00c0}, {0x482b000, 0xc004bdab40}, 0xc006a67aa0, 0xc006a67b40)
	/workspace/lib/limiter/limiter.go:149 +0x2da
github.com/gravitational/teleport/lib/utils.ChainUnaryServerInterceptors.func1.1({0x5a584b0, 0xc008ce00c0}, {0x482b000, 0xc004bdab40})
	/workspace/lib/utils/grpc.go:48 +0x8f
github.com/gravitational/teleport/lib/utils.ErrorConvertUnaryInterceptor({0x5a584b0, 0xc008ce00c0}, {0x482b000, 0xc004bdab40}, 0x3, 0xc006a67ba0)
	/workspace/lib/utils/grpc.go:27 +0x5c
github.com/gravitational/teleport/lib/utils.ChainUnaryServerInterceptors.func1.1({0x5a584b0, 0xc008ce00c0}, {0x482b000, 0xc004bdab40})
	/workspace/lib/utils/grpc.go:48 +0x8f
github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors.UnaryServerInterceptor.func1({0x5a584b0, 0xc008ce00c0}, {0x482b000, 0xc004bdab40}, 0xc006a67aa0, 0xc006a67bc0)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware/v2@v2.0.0-rc.2.0.20220308023801-e4a6915ea237/interceptors/server.go:22 +0x2e4
github.com/gravitational/teleport/lib/utils.ChainUnaryServerInterceptors.func1({0x5a584b0, 0xc008ce00c0}, {0x482b000, 0xc004bdab40}, 0xc006a67aa0, 0xc009016ca8)
	/workspace/lib/utils/grpc.go:52 +0x21f
google.golang.org/grpc.chainUnaryInterceptors.func1.1({0x5a584b0, 0xc008ce00c0}, {0x482b000, 0xc004bdab40})
	/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:1116 +0x126
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.UnaryServerInterceptor.func1({0x5a584b0, 0xc002508360}, {0x482b000, 0xc004bdab40}, 0xc006a67aa0, 0xc0041a43c0)
	/go/pkg/mod/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.31.0/interceptor.go:325 +0x797
google.golang.org/grpc.chainUnaryInterceptors.func1.1({0x5a584b0, 0xc002508360}, {0x482b000, 0xc004bdab40})
	/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:1119 +0x227
google.golang.org/grpc.chainUnaryInterceptors.func1({0x5a584b0, 0xc002508360}, {0x482b000, 0xc004bdab40}, 0xc006a67aa0, 0xc009016ca8)
	/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:1121 +0x286
github.com/gravitational/teleport/api/client/proto._AuthService_GetEvents_Handler({0x48d81c0, 0xc00411fcb0}, {0x5a584b0, 0xc002508360}, 0xc008cca8a0, 0xc006e47820)
	/workspace/api/client/proto/authservice.pb.go:16445 +0x1db
google.golang.org/grpc.(*Server).processUnaryRPC(0xc001bc7880, {0x5a9fca0, 0xc0054d96c0}, 0xc002b51e60, 0xc003479cb0, 0x75c9588, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:1282 +0x1532
google.golang.org/grpc.(*Server).handleStream(0xc001bc7880, {0x5a9fca0, 0xc0054d96c0}, 0xc002b51e60, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:1619 +0xfd4
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:921 +0xfe
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:919 +0x4d1
FAIL	github.com/gravitational/teleport/lib/web	130.579s
```

**Comment by xacrimon** [comment]
@Tener I'd love to have reliable tests where I can query against AWS or use an emulator to test this. The issue is that AWS is at best inconsistent here and I've observed multiple variations of populated fields in cases that aren't really distinct which worries me. Same goes for the AWS emulator, which populates them all regardless as one would expect.

**Comment by xacrimon** [comment]
I'll add some test checks for the dynamic.go changes though. I could potentially construct some nil-check tests for AWS but they seem of dubious value to me since I hardly know what behaviour to expect from AWS.

File: lib/events/dynamoevents/dynamoevents.go [hunk 1]
```diff
@@ -1067,14 +1067,28 @@ func (l *Log) getTableStatus(ctx context.Context, tableName string) (tableStatus
...
 		return false, trace.Wrap(err)
 	}
 
-	return *table.Table.BillingModeSummary.BillingMode == dynamodb.BillingModeProvisioned, nil
+	// Guaranteed to be set.
+	table := res.Table
+
+	// Perform pessimistic nil-checks, assume the table is provisioned if they are true.
+	// Otherwise, actually check the billing mode.
+	switch {
```
Comment (Tener) [inline lib/events/dynamoevents/dynamoevents.go:1082]: I think we can simplify this switch down to a single boolean expression?

```suggestion
	return table.BillingModeSummary == nil || 
	table.BillingModeSummary.BillingMode == nil 
	|| *table.BillingModeSummary.BillingMode == dynamodb.BillingModeProvisioned, nil
```

Comment (xacrimon) [reply lib/events/dynamoevents/dynamoevents.go]: Yep, seems reasonable.

File: lib/events/dynamic_test.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2021 Gravitational, Inc.
```
Comment (Tener) [inline lib/events/dynamic_test.go:2]: Sad but true...

```suggestion
Copyright 2022 Gravitational, Inc.
```

**Review by Tener — APPROVED** [review]

**Comment by xacrimon** [comment]
> Please try running this against a real dynamo instance in both billing modes before merging.

@zmb3 Yep, done. Did that previously as well. Issue was that AWS API is inconsistent in what fields it populates and contains zero docs on when it decides to populate certain fields.

File: lib/events/dynamic.go [hunk 3]
```diff
@@ -37,7 +47,16 @@ func FromEventFields(fields EventFields) (events.AuditEvent, error) {
...
 
-	eventType := fields.GetString(EventType)
+	getFieldEmpty := func(field string) string {
+		i, ok := fields[field]
+		if !ok || !isInterfaceString(i) {
+			return ""
+		}
+
+		return i.(string)
+	}
```
Comment (zmb3) [inline lib/events/dynamic.go:57]: If you're going to do `i.(string)` anyway, you don't need a type switch.

```suggestion
	getFieldEmpty := func(field string) string {
		i, ok := fields[field]
		if !ok {
			return ""
		}
		s, ok := i.(string)
		if !ok {
			return ""
		}

		return s
	}
```

File: lib/events/dynamic_test.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+// TestDynamicTypeUnknown checks that we correctly translate known events into the correct proto type.
+func TestDynamicKnownType(t *testing.T) {
+	fields := EventFields{
+		EventType: "print",
+	}
+
+	event, err := FromEventFields(fields)
+	printEvent := event.(*events.SessionPrint)
+	require.NoError(t, err)
```
Comment (zmb3) [inline lib/events/dynamic_test.go:67]: move the error check before the type assertion

**Review by zmb3 — CHANGES_REQUESTED** [review]

File: lib/events/dynamic.go (hunk 3 above)
Comment (zmb3) [reply lib/events/dynamic.go:57]: Actually, you really can do:
```
s, _ := i.(string)
return s
```

**Review by zmb3 — APPROVED** [review]

**Pull request merged**